src/toxcore/dht/dht_node.rs

```diff
@@ -154,13 +154,44 @@ impl DhtNode {
                         None => Some(SocketAddr::V6(v6)),
                     }
                 },
-                None => {
-                    self.assoc4.saddr.map(Into::into)
-                },
+                None => self.assoc4.saddr.map(Into::into),
             }
         } else {
             self.assoc4.saddr.map(Into::into)
+        };
+
+        if addr.is_none() {
+            warn!("get_socket_addr: failed to get address of DhtNode");
         }
+
+        addr
+    }
+
+    /// Returns all available socket addresses of DhtNode
+    pub fn get_all_sockets(&self, is_ipv6_enabled: bool) -> Vec<SocketAddr> {
+        let mut addrs = Vec::new();
+
+        if is_ipv6_enabled {// DHT node is running in ipv6 mode
+            self.assoc6.saddr.and_then(|v6| Some(addrs.push(SocketAddr::V6(v6))));
+            self.assoc4.saddr.and_then(|v4| {
+                let ip_v6 = v4.ip().to_ipv6_mapped();
+                Some(addrs.push(SocketAddr::new(IpAddr::V6(ip_v6), v4.port())))
+            });
+        } else { // DHT node is running in ipv4 mode
+            self.assoc4.saddr.and_then(|v4| Some(addrs.push(SocketAddr::V4(v4))));
```

Comment:
`if let` is better here since you don't care of the result of `and_then` and mutate `addrs` inside.

Reply:
why?

Reply:
See this line:
```rust
self.assoc4.saddr.and_then(|v4| Some(addrs.push(SocketAddr::V4(v4))));
```
First, we ignore the result of `and_then` which returns `Option` itself.
Second, we artificially pack `addrs.push(SocketAddr::V4(v4))` in `Some` inside so we get `Option<()>` which is kind of ugly.
Third, all we do inside is mutating `addrs` variable so this `and_then` is not clean.

Aren't enough reasons to use simple `if let` construction? :)

Reply:
If you want it in functional style then it should be like:
```rust
self.assoc4.iter().map(SocketAddr::V4)
    .chain(self.assoc6.iter().map(SocketAddr::V6))
    .collect()
```
But not the current approach.

Reply:
OK, I agree